### ISSUE TITLE
[PR Reference] More detailed W&B logs + tensor metrics + loss refactor

### ISSUE THREAD

**Issue by justusmattern27** [Proximal-Labs/prime-rl-clone-5#28]
*This is a reference to original PR #680 by @mikasenghaas*

**Original PR: https://github.com/PrimeIntellect-ai/prime-rl/pull/680**
**Status: closed**
**Created at: 2025-08-05 18:42:04+00:00**
**Merged: True**

This PR refactors some of the W&B logs, especially from the trainer, to have a canonical way to log diverse point statistics (min/max/mean) from tensors (like `probs`, `old_probs`, `importance_ratio`, `entropy`, ...) for more observability during runs. To do this properly, it also required a slight refactor of the loss interface.

- Introduces the `Tensors` class with is used to accumulate local tensors and on a call to `compute_stats` all-gathers the tensors across ranks and steps to compute point metrics (min, max, mean, median, std)
- Refactors input and output of `compute_loss`. It now takes `logprobs` (computed in the main trainer loop), `old_logprobs` as arguments, computes and returns the per-token loss (to be reduced in main trainer loop), as well as a flexible `loss_tensors` dictionary which exports additional tensors for logging (e.g. `importance_ratio`, `clipped_importance_ratio` etc.) which are accumulated into `tensors: Tensors`
- Removed `compute_logprobs` function, instead use `forward`, `shift_logits` and `select_log_softmax` in main loop for logprob recomputation and grad acc loop
- More informative debug micro step logs

Also adds some extra and updates logging keys

- Removes `train`, `orchestrator`, `infer` from any logging keys (e.g.  (e.g. `perf/train/throughput` and `perf/infer/throughput` -> `perf/throughput`); we really don't need them - it's actually nicer to see the same throughput from trainer and orchestrator in one plot
- Adds logging for prompt and completion tokens with keys `prompt_len` and `completion_len`; sequence length is now `seq_len`. Logs `min`, `max`, `mean` and `std` (over full batch)
- Adds logging for `reward` and `advantage` including `min`, `mean` and `max` (over full batch)

## Overview

[W&B View](https://wandb.ai/primeintellect/mika-testing/workspace?nw=10dic2utkcp) with new logging keys.

## Hendryck's

*Run from #688 which is baed on #685 which is based on this PR*

<img width="290" height="250" alt="Screenshot 2025-08-07 at 4 55 59 PM" src="https://github.com/user-attachments/assets/08111afe-db39-48ae-81f1-781c5ede7260" />

## TODOs

- [x] Implement computing and logging tensor point estimates in a standardized way
- [x] We do not have `train`/ `infer` information in any of our keys, e.g. if there is a duplicate key both the trainer and orchestrator will simply log to that key (e.g. `perf/train/throughput` and `perf/infer/throughput` -> `perf/throughput`)
- [x] We add logging for prompt and completion tokens with keys `prompt_len` and `completion_len`
- [x] Check equivalence with 1 vs 2 trainer
- [x] Remove `sum` and `numel` from logging
- [x] Accumulate tensors on CPU and all-gather, then reduce on master rank
- [x] Use `batch`, `seq` and `vocab` for jaxtyping across repo
